File: assignment-4/coding2.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char  magic_number_1[] = {0xBA, 0x5E, 0xBA, 0x11, 0x00};
char  magic_number_2[] = {0xBA, 0x5E, 0xBA, 0x12, 0x00};

// Node structure for linked list
typedef struct Node Node;
struct Node {
    char *word;
    int index;
    Node *next;
};

// String structure consisting of char array, dynamic
struct String {
    int nval;
    int max;
    char *wrd;
} word; // Global variable, word

void check_magic_numbers(FILE *);
char *string_duplicator(char *);
int findnum(int, FILE *);
int add(char);
int del(char);
void write_pos(Node *, int, FILE *, int);
void *emalloc(size_t);
Node *newnode(char *, int);
void print_word(char *, FILE *);
Node *move_to_top(Node *, int, int);
Node *append(Node *, Node *);
void freeall(Node *);
Node *delete(Node *, char *);

/*
 * The encode() function may remain unchanged for A#4.
 */

int encode(FILE *input, FILE *output) {
    return 0;
}

int decode(FILE *input, FILE *output) {
    // Check for magic numbers
    check_magic_numbers(input);
    
    // Initalizing variables
    int ch=0, status=1, num, num_words=0, pos, wc=0;
    Node *dictionary=NULL;

    // Infinite loop
    for(;;) {
        // Newline handling
        if(ch=='\n') {
            fputc('\n', output);
            status=1;
        }

        // Read next char if last word was followed by \n
        if(status) {
            ch=fgetc(input);
            status=0;
        }

        // Break at the end of file
        if(ch==EOF) break;

        // Check if byte is not a letter character
        if(ch>128) {
            // num stores decoded int of char read by file
            num=findnum(ch, input);
            
            // Reads next byte to look ahead for \n
            ch=fgetc(input);

            // Check if num has been encountered before i.e. old word
            if(num<=num_words) {
                pos=num_words-num;
                write_pos(dictionary, pos, output, ch);
                dictionary=move_to_top(dictionary, pos, num_words);
                continue;
            }

            // Loops through characters until a number or \n is encountered and stores it in word
            while(ch<128) {
                add(ch);
                wc++;
                ch=fgetc(input);
                if(ch=='\n') {
                    status=1;
                    break;
                }
            }
            wc=0;

            // Adds new word to dictionary and prints it
            Node *temp=newnode(word.wrd, num_words);
            dictionary=append(dictionary, temp);
            num_words++;
            write_pos(dictionary, num_words-1, output, ch); // Print function
            word.wrd=NULL; // Reset word
        }
    }

    free(word.wrd);
    freeall(dictionary);
    return 0;
}

// Function to check start of file for either of the precious magical number sequences
void check_magic_numbers(FILE *input) {
    int i;
    char ch;
    for(i=0; i<4; i++) {
        ch=fgetc(input);
        if(!(ch==magic_number_1[i] || ch==magic_number_2[i])) {
            fprintf(stderr, "Magic numbers not detected.");
            exit(1);
        }
    }
}

// Returns decoded value of char based on encoding scheme specified in the problem description
int findnum(int ch, FILE *input) {
    int num;
    num=ch-128;
    if(num==121) {
        ch=fgetc(input);
        num=ch+121;
    }
    else if(num==122) {
        ch=fgetc(input);
        num=(ch*256)+376;
        ch=fgetc(input);
        num+=(ch%256);
    }
    return num;
}

// Very slightly modified version of addname() function taken from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG265 Resources Slide 27
int add(char chr) {
    char *str;
    if(word.wrd==NULL) {
        word.wrd=(char *)malloc(1*sizeof(char));
        if(word.wrd==NULL) {
            return -1;
        }
        word.max=1;
        word.nval=0;
    }
    else if(word.nval>=word.max) {
        str=(char *)realloc(word.wrd, (2*word.max)*sizeof(char));
        if(str==NULL) {
            return -1;
        }
        word.max=2*word.max;
        word.wrd=str;
    }
    word.wrd[word.nval]=chr;
    return word.nval++;
}

// Very slightly modified version of delname() function taken from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG265 Resources Slide 29
int del(char chr) {
    int i;
    for(i=0; i<word.nval; i++) {
        if(word.wrd[i]==chr) {
            memmove(word.wrd+i, word.wrd+i+1, (word.nval-(i+1))*sizeof(char));
            word.nval--;
            return 1;
        }
    }
    return 0;
}

// string_duplicator() taken from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG265 resources Slide 30
char *string_duplicator(char *input) {
    char *copy;
    assert(input!=NULL);
    copy=(char *)emalloc(sizeof(char)*strlen(input)+1);
    strncpy(copy, input, strlen(input)+1);
    return copy;
}

// Function to print word corresponding to position, pos
void write_pos(Node *dictionary, int pos, FILE *output, int ch) {
    Node *p;
    char *word;
    for(p=dictionary; p!=NULL; p=p->next) {
        if(p->index==pos) {
            word=p->word;
            break;
        }
    }
    if(ch=='\n') {
        print_word(word, output);
    }
    else {
        print_word(word, output);
        fputc(' ', output);
    }
}

// Simple char array print function
void print_word(char *word, FILE *output) {
    int i;
    for(i=0; i<strlen(word); i++) {
        fputc(word[i], output);
    }
}

// Moves node corresponding to position, pos to the top of the dictionary
Node *move_to_top(Node *dictionary, int pos, int num_words) {
    char *word;
    Node *p;
    for(p=dictionary; p!=NULL; p=p->next) {
        if(p->index==pos) {
            word=p->word;
            break;
        }
    }
    dictionary=delete(dictionary, word);
    num_words--;
    int i=0;
    for(p=dictionary; p!=NULL; p=p->next) {
        p->index=i;
        i++;
    }
    Node *temp=newnode(word, num_words);
    dictionary=append(dictionary, temp);
    num_words++;
    return dictionary;
}

// emalloc() function taken from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG 265 resources Slide 38
void *emalloc(size_t n) {
    void *m;
    m=malloc(n);
    if(m==NULL) {
        fprintf(stderr, "malloc of %u bytes failed", n);
        exit(1);
    }
    return m;
}

// newnode() function taken from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG265 Resources Slide 39 
Node *newnode(char *word, int index) {
    Node *new;
    new=(Node *)emalloc(sizeof(Node));
    new->word=word;
    new->index=index;
    new->next=NULL;
    return new;
}

// append function taken from addend() function from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG265 Resources Slide 41
Node *append(Node *list, Node *new) {
    Node *p;
    if(list==NULL) {
        return new;
    }
    for(p=list; p->next != NULL; p=p->next);
    p->next=new;
    return list;
}

// delete() taken from delitem() function from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG265 Resources Slide 50
Node *delete(Node *list, char *word) {
    Node *curr, *prev;
    prev=NULL;
    for(curr=list; curr!=NULL; curr=curr->next) {
        if(strcmp(word, curr->word)==0) {
            if(prev==NULL) {
                list=curr->next;
            }
            else {
                prev->next=curr->next;
            }
            free(curr);
            return list;
        }
        prev=curr;
    }
    fprintf(stderr, "node %s not in list", word);
    exit(1);
}

// freeall() function taken from 06_c_dynamic_memory.pdf on connex.csc.uvic.ca SENG265 Resources Slide 48
void freeall(Node *list) {
    Node *next;
    for(; list!=NULL; list=next) {
        next=list->next;
        free(list);
    }
}

File: assignment-3/coding2.py
#!/usr/bin/python3 

import os
import sys
import re

MAGIC_NUMBER_1 = chr(0xBA) + chr(0x5E) + chr(0xBA) + chr(0x11)
MAGIC_NUMBER_2 = chr(0xBA) + chr(0x5E) + chr(0xBA) + chr(0x12)

class Files:
	"""A class defining the infile and outfile"""
	def __init__(self):
		self.infile=None
		self.outfile=None

	# For opening files with encoder
	def eopen(self, input_name, output_name):
		try:
			self.infile=open(input_name, 'r')
			self.outfile=open(output_name, encoding='latin-1', mode='w')
		except:
			print("Error. Exception was raised when opening the file.")
			sys.exit(1)

	# For opening files with decoder
	def dopen(self, input_name, output_name):
		try:
			self.infile=open(input_name, encoding='latin-1', mode='r', newline="")
			self.outfile=open(output_name, 'w')
		except:
			print("Error. Exception was raised when opening the file.")
			sys.exit(1)

	# Close all files
	def fclose(self):
		self.infile.close()
		self.outfile.close()

def encode(input_name):
	(base_name, _, _) = input_name.rpartition(".")
	output_name = base_name + "." + "mtf"

	# Creating instance of class Files
	files=Files()
	files.eopen(input_name, output_name)

	# Writes magic number sequence to outfile
	write_magic_numbers(files)

	# Initialzing variables
	dictionary=[]
	status=1

	# Using re to split entire file into a list of words and '\n' characters
	text=files.infile.read()
	regex=re.split(r'(\s?|\n)', text)
	for item in regex:
		if item==' ' or item=='':
			regex.pop(regex.index(item))

	# Cycle through words in the list returned by re
	for word in regex:
		# Newline handling (& empty line handling)
		if word=='\n':
			files.outfile.write('\n')
			continue

		# Cycle through words in dictionary to attempt to find word it has encountered before
		for d_word in dictionary:
			if word==d_word:
				# Old word handling
				status=0
				index=dictionary.index(d_word)
				pos=len(dictionary)-index
				write_code_to_outfile(pos, files)
				# Rearraging dictionary
				move_to_top(d_word, index, dictionary)
				break
		if status:
			# New word handling
			dictionary.append(word)
			write_code_to_outfile(len(dictionary), files)
			files.outfile.write(word)
		status=1

	# Closing files
	files.fclose()

# Writes the magic number sequence to the file
def write_magic_numbers(files):
	global MAGIC_NUMBER_2
	for code in MAGIC_NUMBER_2:
		files.outfile.write(code)

# Moves current d_word to the top of the dictionary
def move_to_top(d_word, index, dictionary):
	dictionary.pop(index)
	dictionary.append(d_word)

# Outputs the right chr based on the value of code
def write_code_to_outfile(code, files):
	if code>=1 and code<121:
		files.outfile.write(chr(code+128))
	elif code>120 and code<376:
		files.outfile.write(chr(121+128)+chr(code-121))
	elif code>375 and code<=65912:
		files.outfile.write(chr(122+128)+chr((code-376)//256)+chr((code-376)%256))

def decode(input_name):
	(base_name, _, _) = input_name.rpartition(".")
	output_name = base_name + "." + "txt"

	# Creating instance of class Files
	files=Files()
	files.dopen(input_name, output_name)

	# Checking for mtf sequences
	check_magic_numbers(files)

	# Initializing variables
	dictionary=[]
	word=""
	status=1
	byte=None

	# Loops through chars in the file
	while True:
		# Newline handling
		if byte=='\n':
			files.outfile.write(byte)
			status=1

		# Read next byte if last word was terminated due to \n
		if status:
			byte=files.infile.read(1)
			status=0

		# Break at the end of the file
		if not byte:
			break

		# Check if byte is not a character
		if ord(byte)>128:
			# num holds decoded int value of num byte
			num=findnum(files, byte)

			# Reads next byte (sorta looking ahead to check for \n)
			byte=files.infile.read(1)

			# Check if the number has been encountered before i.e. old word
			if num<=len(dictionary):
				pos=len(dictionary)-num
				write_pos(pos, dictionary, files, byte)
				# Rearrange dictionary
				d_move_to_top(dictionary, pos)
				continue

			# Loops through chars until a number or '\n' is encountered and stores it in word
			while ord(byte)<128:
				word+=byte
				byte=files.infile.read(1)
				if byte=='\n':
					status=1
					break

			# Adds new word to dictionary and prints it
			dictionary.append(word)
			write_pos(-1, dictionary, files, byte)
			word=""

	# Closing files
	files.fclose()

# Checks if the file has either of the 2 magic number sequences in it
def check_magic_numbers(files):
	global MAGIC_NUMBER_1, MAGIC_NUMBER_2
	MAGIC_NUMBER_1=list(MAGIC_NUMBER_1)
	MAGIC_NUMBER_2=list(MAGIC_NUMBER_2)
	m_nums=[]
	for i in range(4):
		m_nums.append(chr(ord(files.infile.read(1))))
	if not (m_nums==MAGIC_NUMBER_1 or MAGIC_NUMBER_2):
		print("Error. Magic numbers not detected.")
		sys.exit(1)

# Moves word at position, pos to the top of the dictionary
def d_move_to_top(dictionary, pos):
	temp=dictionary[pos]
	dictionary.pop(pos)
	dictionary.append(temp)

# Returns decoded int value of byte
def findnum(files, byte):
	num=ord(byte)-128
	if num==121:
		byte=files.infile.read(1)
		num=ord(byte)+121
	elif num==122:
		byte=files.infile.read(1)
		num=(ord(byte)*256)+376
		byte=files.infile.read(1)
		num+=(ord(byte)%256)
	return num

def write_pos(pos, dictionary, files, byte):
	if byte=='\n':
		files.outfile.write(dictionary[pos])
	else:
		files.outfile.write("%s " % dictionary[pos])
